Project 5 For John Beverly

```diff
@@ -51,3 +51,441 @@ You may find the following resources useful:
 The class will be divided evenly. Students are expected to determine allotment, but teams must be comprised of the same number of members (with a +1/-1 deviation allowed). This is an exercise in project management. I encourage you to identify a strategy early for dividing sub-tasks, setting deadlines, and addressing blockers to progress.  
 
 I suggest setting up 15 minute 'stand-up' meetings every other day, where team members are expected to join. zoom call or meet in person and (a) explain what progress they have made on their sub-task, (b) explain what they intend to achieve before the next stand-up, and (c) share any blockers that have to progress. This is an effective way to keep members focused on a team goal, as well as opportunities to overcome challenges. 
+
+
+
+
+Group 2 contains Emanuel, Katie, Josh V (commonly know as V by group member), Cameron, Josh B, Jieming, Lance, Tim(?), also Delaney 
+Please note we DID check this against SHACL playground :) 
+enjoy, if possible.
+
+1.	Constraint for Asymmetry 
+
+a.	Our constraint checks for asymmetric patterns in a graph.
+
+b.	An example of when to use this constraint would be for tracking enrollment in courses at a university. Such that each node would represent a student of course and each relation represents the enrollment of a student in a course. Enrollment relations should only be allowed between students and courses that are in a department. If a violation of this constraint is detected it would mean that there is an enrollment relation that does not satisfy the requirement of a student enrolling in a course without the course being enrolled in a student. By changing ex:AsymmteryShape to ex:AsymmetryEnrollementShape and following our basic structure. 
+
+c.	(a -> b) ^ ~(b -> a)
+
+@prefix sh: <http://www.w3.org/ns/shacl#> .
+@prefix ex: <http://example.com/> .
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
+
+
+ex:AsymmetryShape
+
+a sh:NodeShape ;
+
+sh:targetClass ex:relation ;
```

Comment:
Sure this is just a basic framework. The goal was for any person to grab these constraints and use them for their needs so there is a possibility one could use this constraint to check for proper parthood. Because our constraints were not supposed to be tied to any specific ontology we just want them to be general enough that anybody could utilize them in their ontologies.

Reply:
Let me put another way :) It seems that the purpose of your shape is to check whether object properties in any ontology are asymmetric. But in order to make shape work, you also have to treat those object properties as relation classes in that ontology. Unless a ontology includes not only object properties like R but also relation classes R-ness, your shape cannot work as it intends. **So my real concern is about the use of "sh:targetClass".** I think you don't need to understand a relation or an object property as a class. You can still understand it as a relation by using "targetSubjectsOf" or "targetObjectsOf" to add some contraints to its relata.

Reply:
@peihongx  how does this look?

@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix ex: <http://example.com/> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .

ex:SymmetricRelationShape
    a sh:NodeShape ;
    sh:targetClass ex:Relation ;
    sh:property [       
 sh:path ex:hasSource ;    
] ;
    sh:property [        
sh:path ex:hasTarget ;        
sh:minCount 1 ;        
sh:maxCount 1 ;        
sh:message "A relation must have exactly one target"   
 ] ;
    sh:property [        
sh:path ex:hasInverseRelation ;       
 sh:minCount 0 ;        
sh:maxCount 1 ;        
sh:class ex:Relation ;        
sh:node [           
 sh:property [                
sh:path ex:hasSource ;            
] ;
            sh:property [                
sh:path ex:hasTarget ;                
sh:minCount 1 ;                
sh:maxCount 1 ;                
sh:message "An inverse relation must have exactly one target"            
] ;
            sh:property [                
sh:path ex:hasInverseRelation ;                
sh:minCount 0 ;                
sh:maxCount 1 ;                
sh:class ex:Relation ;                
sh:node [                   
 sh:property [                        
sh:path ex:hasSource ;                        
sh:equals [ 
sh:path [ 
rdf:reverseProperty ex:hasTarget 
]
 ]
] ;
                    sh:property [                        
sh:path ex:hasTarget ;                        
sh:equals [ 
sh:path [ 
rdf:reverseProperty ex:hasSource 
] 
]
                    ]
                ]
            ] ;
            sh:property [               
 sh:path ex:hasSource ;                
sh:equals [
 sh:path [ 
rdf:reverseProperty ex:hasTarget 
] 
]
            ] ;
            sh:property [                
sh:path ex:hasTarget ;                
sh:equals [ 
sh:path [ 
rdf:reverseProperty ex:hasSource 
] 
]
            ]
        ]
    ] .

Reply:
Happy to know your progress! You know, any novel things are always difficult to propose and defend. Your team is doing something unusual! 👍

Reply:
I am not sure if this will work but I hope it can.

Reply:
sh:property [
sh:path ex:hasInverseRelation ;
sh:minCount 0 ;
sh:maxCount 1 ;
sh:class ex:Relation ;
sh:node [
sh:property [
sh:path ex:hasSource ;
] ;

**Maybe we should use sh:inversePath instead of ex:hasInverseRelation? Similarly, I am wondering whether rdf:reverseProperty can be changed to sh:inversePath** 

sh:property [
sh:path ex:hasTarget ;
sh:minCount 1 ;
sh:maxCount 1 ;
sh:message "An inverse relation must have exactly one target"
]

This property shape still applies to the original target class, so I'm afraid that what it means is that a relation (but not an inverse relation, as you intend) must have exactly one target.

BTW, I'm not very sure why we wanna add constraints about source or the number of targets.